File: 프로그래머스/2/42584. 주식가격/주식가격.py
def solution(prices):
    answer = [0] * len(prices)
    stack = []
    
    for i in range(len(prices)):
        while stack and prices[stack[-1]] > prices[i]:
            answer[stack[-1]] = i - stack[-1]
            stack.pop()
        stack.append(i)
        
    for s in stack:
        answer[s] = len(prices) - s - 1
        
    return answer

File: 프로그래머스/2/42584. 주식가격/주식가격.java
import java.util.*;

class Solution {
    public List<Integer> solution(int[] prices) {
        List<Integer> answer = new ArrayList<>();
        for (int i = 0; i < prices.length; i++) {
            answer.add(0);
        }
        
        Stack<Integer> stack = new Stack<>();
        for (int i = 0; i < prices.length; i++) {
            while (!stack.isEmpty() && prices[stack.peek()] > prices[i]) {
                int idx = stack.pop();
                answer.set(idx, i - idx);
            }
            stack.push(i);
        }
        
        while (!stack.isEmpty()) {
            int idx = stack.pop();
            answer.set(idx, prices.length - idx - 1);
        }
        return answer;
    }
}